src/web/js/output-ui.js

```diff
@@ -1159,92 +877,63 @@
-              
-              var locClasses = cmLocs.map(
-                function(l){return cmlocToCSSClass(l);});
-              
-              //for (var h = 0; h < locArray.length; h++) {
-              //  anchor.addClass(locClasses[h]);
-              //  var highlight = highlights.get({pl:locArray[h],l:cmLocs[h],c:cssColor});
-              //  if(highlight == undefined) {
-              //    highlights.set({pl:locArray[h],l:cmLocs[h],c:cssColor},
-              //      highlightSrcloc(runtime, editors, srcloc, locArray[h], cssColor, context, true));
-              //  }
-              //}
-              for (var h = 0; h < locArray.length; h++) {
-                anchor.addClass(locClasses[h]);
-                var highlight = highlights.get({pl: locArray[h], l: cmLocs[h], c: cssColor});
-                if(highlight == undefined) {
-                  highlights.set({pl: locArray[h], l: cmLocs[h], c: cssColor},
-                                 highlightSrcloc(runtime, editors, srcloc, locArray[h], cssColor, context, true));
-                }
-              }
-              
-              if (hue === undefined) {
-                anchor.on("mouseenter", function() {
-                  for (var i = 0; i < locClasses.length; i++) {
-                    hintLoc(runtime, editors, srcloc, locArray[i]);
-                    $("."+locClasses[i]).css("animation", "pulse-underline 0.4s infinite alternate");
-                  }
-                });
-              } else {
-                anchor.on("mouseenter", function() {
-                  for (var i = 0; i < locClasses.length; i++) {
-                    hintLoc(runtime, editors, srcloc, locArray[i]);
-                    $("."+locClasses[i]).css("animation", "pulse 0.4s infinite alternate");
-                  }
+              var positions = ffi.toArray(locs).
+                filter(isSrcloc).
+                map(function(loc){
+                  return Position.fromPyretSrcloc(runtime, srcloc, loc, documents);
                 });
+              if (id < 0) {
+                messageHintedColors.add(color);
               }
-              
-              anchor.on("click", function() {
-                for (var z = 0; z < locClasses.length; z++) {
-                  var cmloc = cmLocs[z];
-                  var els = document.getElementsByClassName(locClasses[z]);
-                  emphasizeLine(editors, cmloc);
-                }
+              if(!messageAnchors.has(color))
+                messageAnchors.set(color, [anchor]);
+              else messageAnchors.get(color).push(anchor);
+              if(!messagePositions.has(color))
+                messagePositions.set(color, positions);
+              else Array.prototype.push.apply(messagePositions.get(color), 
+                                              positions);
+              anchor.on("click", function (e) {
+                if (positions[0] !== undefined)
```

Comment:
Is this check catching the case where `positions` could be _empty_?   It doesn't seem like `fromPyretSrcloc` can return `undefined`, so should this be a check for `length`?  Or should the `"click"` event simply not be registered if `positions` is empty?